[optim] Merge the pyi files into py files of optimizer

david20571015 opened the pull request:
Continue the work of pytorch/pytorch#125153

david20571015 committed:
typing: adam

david20571015 committed:
typing: adamax

david20571015 committed:
typing: adamw

david20571015 committed:
typing: asgd

david20571015 committed:
typing: lbfgs

david20571015 committed:
typing: nadam

david20571015 committed:
typing: radam

david20571015 committed:
typing: rmsprop

david20571015 committed:
typing: rprop

david20571015 committed:
typing: sparse_Adam

david20571015 committed:
fix: import of TypeAlias

david20571015 committed:
typing: swa_utils

david20571015 committed:
style: modify the variable name to fit other optimizer's naming style

david20571015 committed:
typing: sgd

david20571015 committed:
init

david20571015 commented:
@pytorchbot drci

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -708,16 +716,17 @@ def _fused_adam(
-    lr_dict = (
-        {lr.device: lr} if isinstance(lr, Tensor) and str(lr.device) != "cpu" else None
-    )
+    lr = torch.tensor(lr)
```
why is this necessary? we don’t want to wrap lr into a Tensor normally

david20571015 replied:
I believe my modification here may be incorrect, I will fix this.

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -27,10 +29,10 @@ class AdamW(Optimizer):
         self,
         params: ParamsT,
-        lr: Union[float, Tensor] = 1e-3,
+        lr=1e-3,
```
why the removal of types here?

david20571015 replied:
reverted
I accidently removed it.

janeyx99 commented on torch/optim/asgd.py:
```diff
@@ -376,6 +377,7 @@ def _multi_tensor_asgd(
             torch._foreach_add_(grouped_state_steps, 1)
 
         # intermediate = grad + param * lambd
+        intermediate: Union[Tuple[Tensor, ...], List[Tensor]]
```
what’s the diff between defining this here vs the first time it’s instantiated?

david20571015 replied:
without this, mypy will raise an error `Incompatible types in assignment (expression has type "tuple[Tensor, ...]", variable has type "list[Tensor]")` in line 386, 392 and 411 because `intermediate` was assigned as list[Tensor] in line 384. I think it is more concise to assign a union type than add 3 type ignores.

janeyx99 reviewed (commented):
Started reviewing, but it’d be easier to review if you could split the lr_scheduler changes from the optims

david20571015 committed:
fix: lr type in _fused_adam

david20571015 committed:
fix: lr type in Adamw

david20571015 commented:
thanks for your review @janeyx99, I splited the changes about lr_scheduler into #125556.

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -27,10 +29,10 @@ class Adam(Optimizer):
         self,
         params: ParamsT,
-        lr: Union[float, Tensor] = 1e-3,
+        lr=1e-3,
```
looks like there are more places you accidentally removed

david20571015 replied:
fixed. sorry for this.

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -701,16 +710,17 @@ def _fused_adamw(
-    lr_dict = (
-        {lr.device: lr} if isinstance(lr, Tensor) and str(lr.device) != "cpu" else None
-    )
+    lr = torch.tensor(lr)
```
remove

david20571015 replied:
fixed

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -688,11 +697,11 @@ def _fused_adamw(
     weight_decay: float,
     eps: float,
     maximize: bool,
-    capturable: bool,  # Needed for consistency.
```
why delete the comment?

david20571015 replied:
I thought this is unnecessary because `has_complex` didn't have this comment. But I add it back and add a same comment for `has_complex`.

janeyx99 commented on torch/optim/swa_utils.py:
```diff
@@ -242,15 +238,8 @@ def update_parameters(self, model: Module):
-                        Tuple[List[List[Tensor]], Indices],
-                    ],
-                    grouped_tensors,
+                grouped_tensors = Optimizer._group_tensors_by_device_and_dtype(
```
Why the switch to use the optimizer one here?

david20571015 replied:
The optimizer one internally calls the original function and also supports compiling. It doesn't need to perform a lot of type casting because it is ignored in the optimizer one. Is this not suitable here?"

janeyx99 replied:
eh...the one in optimizer is a special case where it gets skipped in compile world..I'd rather not use it here if not necessary yet.

david20571015 replied:
Okay, I'll roll it back. Thanks for explaining this.

janeyx99 reviewed (commented):
Almost there!

david20571015 committed:
fix: lr type in Adam

david20571015 committed:
fix: fused adamw

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -29,8 +31,8 @@ def __init__(
         betas: Tuple[float, float] = (0.9, 0.999),
-        eps: float = 1e-8,
-        weight_decay: float = 0,
+        eps=1e-8,
+        weight_decay=0,
```
here too

janeyx99 replied:
though I see other places where this wasn't necessary--do you know the difference between defining vs not here?

If it's fine to not have a type, I am happy with not having one.

david20571015 replied:
Sorry, could you please explain the difference to me? I think mypy interprets weight_decay as an int, so should I specify a float here or change 0 to 0.0?

janeyx99 replied:
It does look like there's no error and it will allow an int or a float or anything (it might just infer it as Any). From my lil endeavor, it still feels the best to include the type. For example:

```
class Foo:
    def __init__(self, arg, kwarg=0):
        self.arg = arg
        self.kwarg = kwarg

a = Foo(1, "l")

print(5 + a.kwarg)
```
This will not error and will do the type conversion automatically.

But the intention is closer to: 
```
class Foo:
    def __init__(self, arg, kwarg: int = 0):
        self.arg = arg
        self.kwarg = kwarg

a = Foo(1, "l")

print(5 + a.kwarg)
```
And here, with the type, mypy will error properly at the instantiation of a

david20571015 replied:
Okay, understood completely. I'll go ahead and add the types.

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -29,8 +31,8 @@ def __init__(
         betas: Tuple[float, float] = (0.9, 0.999),
-        eps: float = 1e-8,
-        weight_decay: float = 1e-2,
+        eps=1e-8,
+        weight_decay=1e-2,
```
here

david20571015 committed:
fix: rollback swa_utils

david20571015 committed:
add numerical types

david20571015 commented:
@pytorchbot merge

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -318,92 +318,6 @@ def step(self, closure=None):
 )
 
 
-def adamw(
```
wait is this deletion intended?

david20571015 replied:
I merged main branch to this, and [0f02e0a](https://github.com/pytorch/pytorch/commit/0f02e0aa39b3dff15a8e2d1a81d049e838d7c696) moved the function to the end of the file.

david20571015 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge

david20571015 committed:
fix: del unused module after import

david20571015 commented:
@pytorchbot rebase -b main

david20571015 committed:
fix: typing for torchscript

david20571015 committed:
unite the usage of `DeviceDict`

david20571015 commented:
@pytorchbot merge

david20571015 commented:
Sorry @janeyx99, I mistakenly thought the failures might been fixed, so I rebase to the main branch. However, after digging into the error message, I found that was caused by my code and I fixed it in https://github.com/pytorch/pytorch/pull/125452/commits/165fed5261170b85f70ae4e64365b7e77a0cea68. Could you approve the CI again please.

Sorry @janeyx99, I mistakenly thought the failures might have been fixed, so I rebased onto the main branch. However, after investigating the error message, I found that it was caused by my code, and I fixed it in https://github.com/pytorch/pytorch/pull/125452/commits/165fed5261170b85f70ae4e64365b7e77a0cea68. Could you please approve the CI again?

david20571015 commented:
Should I keep the parameter name `d_p_list`? None of the affected functions are exposed to external users.

![image](https://github.com/pytorch/pytorch/assets/51911434/615e5acd-021f-44e1-83c4-a8b6318365a3)

janeyx99 commented:
> Should I keep the parameter name `d_p_list`? None of the affected functions are exposed to external users.

Yes, you should keep the name. sgd is publicly exposed and the argument is sadly not positional only.

janeyx99 commented:
The removal in optimizer `__init__.py` is the torch.optim.sgd module, not the function, which is offered so people can manage their own state. (I have yet to figure out why we do the dels there but that is not relevant for this discussion.)

david20571015 committed:
revert d_p_list for BC

david20571015 commented:
> > Should I keep the parameter name `d_p_list`? None of the affected functions are exposed to external users.
> 
> Yes, we should keep the name. sgd is publicly exposed and the argument is sadly not positional only.

done.

janeyx99 commented:
looks like there are merge conflicts now :/

david20571015 committed:
Merge branch main into patch

david20571015 commented:
> looks like there are merge conflicts now :/

fixed

janeyx99 commented:
Please run lintrunner locally to ensure lint passes for the next commit. Getting lintrunner locally is pretty simple: just
```
pip install lintrunner
lintrunner init
lintrunner -a
```

david20571015 committed:
fix: lint

david20571015 commented:
> Please run lintrunner locally to ensure lint passes for the next commit (and usually before pushing). Getting lintrunner locally is pretty simple: just
> 
> ```
> pip install lintrunner
> lintrunner init
> lintrunner -a
> ```

ok! fixed
